Use getrandom(2), if available, to implement Interop.GetRandomBytes()

lpereira opened the pull request:
(These changes are based on similar changes I made in coreclr.)

The getrandom(2) system call, introduced in Linux 3.17, allows one to obtain numbers from the same pool exported as /dev/urandom (if called with the GRND_NONBLOCK flag).  The main advantage is that there's no need to keep a file descriptor open, but this also means that a good source of entropy is always available, even if /dev isn't present (such as in a tight chroot environment).

Refactor how the method to obtain random numbers by using an ifunc, which will continue to prefer arc4random_buf(), if available, and pick between getrandom(2) or reading from /dev/urandom.  Falling back to lrand48(3) happens if either of these last two methods fail for whatever reason.

It's important to note that the entropy quality for /dev/urandom and whatever getrandom(2) returns is already pretty good, so it might not be a good idea to XOR it with values from lrand48(3).  Since we're not dealing with a CSPRNG here, though, it's good to leave that as the fallback mechanism.

lpereira committed:
O_CLOEXEC should be OR-ed with the flags argument, not mode

lpereira committed:
Use getrandom(2), if available, to implement Interop.GetRandomBytes()

The getrandom(2) system call, introduced in Linux 3.17, allows one to
obtain numbers from the same pool exported as /dev/urandom (if called
with the GRND_NONBLOCK flag).  The main advantage is that there's no
need to keep a file descriptor open, but this also means that a good
source of entropy is always available, even if /dev isn't present (such
as in a tight chroot environment).

Refactor how the method to obtain random numbers by using an ifunc,
which will continue to prefer arc4random_buf(), if available, or pick
between getrandom(2) or reading from /dev/urandom.

The behavior of XOR-ing the results with lrand48(3) is still there as a
workaround broken hardware RNGs.  This PRNG is still initialized with
time(NULL), which doesn't have a lot of entropy, and #38034 has been
filed to address that in the future.

vcsjones commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,206 @@
         {
-            int32_t offset = 0;
-            do
+            if (errno == EINTR)
```
According to the docs, `getrandom` will return -1 and `errno` will be `EAGAIN` when `GRND_NONBLOCK` is used and the entropy pool is not available. Should this include `EAGAIN` to try again?

lpereira replied:
Yes, it's often the case during early boot scenarios, or on embedded systems. However, calling getrandom() in a tight loop if it keeps returning EAGAIN might turn into an infinite loop in such cases; I'm not sure it's advisable.

bartonjs replied:
I have at least one VM that can take minutes to fully initialize.

Ubuntu 16.04.6
```
$ dmesg | grep random
[    1.844973] random: udevadm: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.846829] random: udevadm: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.849528] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.850996] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.852320] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.864731] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.866058] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.867459] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.870948] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.872280] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[  180.310815] random: nonblocking pool is initialized
```

Newer kernels trust the CPU, so it's less likely:

Ubuntu 19.04:
```
$ dmesg | grep random
[    0.017218] random: get_random_bytes called from start_kernel+0x97/0x516 with crng_init=0
[    0.536646] random: crng done (trusting CPU's manufacturer)
```

Though not necessarily:

Ubuntu 18.04:

```
$ dmesg | grep random
[    0.000000] random: get_random_bytes called from start_kernel+0x99/0x4fd with crng_init=0
[    1.452697] random: fast init done
[    1.454713] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[    1.462508] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[    1.466497] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[   10.935718] random: crng init done
[   10.935721] random: 7 urandom warning(s) missed due to ratelimiting
```

So the EAGAIN either needs to fail over (e.g. to "dirty reads" from /dev/urandom), or the read should just be a blocking read.  But it's definitely an important case to handle.

bartonjs replied:
(Possibly it's already being handled OK, just that the assert needs to be removed)

lpereira replied:
The assert was just a copy from the read(/dev/urandom) codepath.  Maybe a good idea to remove it from there.

BTW, not sure if you're able to set this up in your VM, but there is a virtio-rng device that some hypervisors implement (qemu-kvm being one of them) that provide entropy from the host to the guest; the Linux kernel is able to use it if detected.

bartonjs replied:
> How does that jive with it potentially taking minutes to initialize? I don't see how that's viable.

Only at boot.  Once a system has initialized it'll never block again for any process, until shutdown.

/dev/urandom will give "dirty random" (which we currently shrug and hope is adequate, and it probably is), getrandom() will (IIUC) leave the data uninitialized (aka simply not write to it).  My comment is just that we can't leave it uninitialized, so either loop, block, or read /dev/urandom.  Looping and blocking are effectively the same thing.

stephentoub replied:
> so either loop, block, or read /dev/urandom

And my comment is we shouldn't block or loop if we may do so for any measurable period of time.

GrabYourPitchforks replied:
@bartonjs Is the "no entropy available" message you're receiving on a fresh VM, or on a VM that has been rebooted? Per http://man7.org/linux/man-pages/man4/random.4.html:

>  If a seed file is saved across reboots as recommended below (all
       major Linux distributions have done this since 2000 at least), the
       output is cryptographically secure against attackers without local
       root access as soon as it is reloaded in the boot sequence, and
       perfectly adequate for network encryption session keys.

IMO blocking is the preferred behavior, especially if it happens only very rarely. If somebody's calling our APIs expecting high-entropy random values then we need to give them high-entropy random values. We can't pull a sleight of hand and give them low-entropy values.

stephentoub replied:
> IMO blocking is the preferred behavior

This caused significant problems when we were initially bringing up .NET on Linux.  We initially had a blocking implementation, and start-up of the process could end up getting blocked for seconds, until the user wildly moved their mouse around to introduce entropy into the system.

Maybe what we're talking about here is a special case somehow, or maybe things have been improved on all distros/versions we now support, but I don't believe blocking waiting for sufficient entropy is viable.

GrabYourPitchforks replied:
FWIW, the Python crew considered this same problem a few years ago. They took a breaking change between Python 3.5.2 and Python 3.6 in order to make the `os.urandom()` function block on low-entropy scenarios because of the security risks inherent in the original behavior. See https://www.python.org/dev/peps/pep-0524/ for more information.

jkotas replied:
`os.urandom` is returns bytes suitable for cryptographic use (https://docs.python.org/2/library/os.html#os.urandom).

`SystemNative_GetNonCryptographicallySecureRandomBytes` is not meant to be suitable for cryptographic use as the name suggests.

GrabYourPitchforks replied:
If this API isn't intended to produce strong randomness, we should reconsider having `Guid.NewGuid` depend on it. We know from experience that for better or worse, devs use `Guid.NewGuid` to produce random identifiers and expect those identifiers to have strong randomness.

lpereira commented:
The entropy pool might not be available during early boot situations, and busy-retrying on EAGAIN might not be the best idea, as it may be stuck in that loop for a long time.

From: Kevin Jones <notifications@github.com>
Sent: Thursday, May 23, 2019 2:09 PM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Leandro Pereira <Leandro.Pereira@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [dotnet/corefx] Use getrandom(2), if available, to implement Interop.GetRandomBytes() (#37906)


@vcsjones commented on this pull request.

________________________________

In src/Native/Unix/System.Native/pal_random.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F37906%23discussion_r287134978&data=02%7C01%7Cleandro.pereira%40microsoft.com%7Cabd1f07a418046037c6508d6dfc2d992%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636942425303742823&sdata=Kpu5H3BMXwUphd97y0J0NDGm4lUfpVvXHp2C2yNVPDA%3D&reserved=0>:

>          {

-            int32_t offset = 0;

-            do

+            if (errno == EINTR)

According to the docs, getrandom will return -1 and errno will be EAGAIN when GRND_NONBLOCK is used and the entropy pool is not available. Should this include EAGAIN to try again?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F37906%3Femail_source%3Dnotifications%26email_token%3DAAADVGJ7AWPSFBYONJXE7FLPW4BWBA5CNFSM4HPKCJY2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZSEFTY%23pullrequestreview-241451727&data=02%7C01%7Cleandro.pereira%40microsoft.com%7Cabd1f07a418046037c6508d6dfc2d992%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636942425303742823&sdata=Dp7FLDrKWLVPUFS%2BLoMd6rKA2wvT0WwwuG63H7ShgDw%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAADVGI7NEK55VRZOO35553PW4BWBANCNFSM4HPKCJYQ&data=02%7C01%7Cleandro.pereira%40microsoft.com%7Cabd1f07a418046037c6508d6dfc2d992%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636942425303752835&sdata=Z%2F1L5LGffDepcmxXiVYvqeVbcQ3ZHpdVMG0hHRqbOKg%3D&reserved=0>.

stephentoub commented:
cc: @janvorli, @bartonjs

danmosemsft commented:
@tmds

lpereira commented:
(Some of the build break is because this uses the ifunc attribute that's not supported by some of the older compilers used in CI. I'll address this with a different strategy, but I still want input on the approach.)

am11 commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,206 @@
+
+#ifndef GRND_NONBLOCK
+#define GRND_NONBLOCK 0x1
+#endif // GRND_NONBLOCK
```
This definition can be replaced in favor of system header. `GRND_NONBLOCK` requires `#include <sys/random.h>` or `<linux/random.h>`; the former one is preferred to make it work on Solaris as well: https://docs.oracle.com/cd/E88353_01/html/E37841/getentropy-2.html (tested with clang39 and gcc7 on Ubuntu, and gcc7 on SmartOS).

lpereira replied:
Done. Also checking for the existence of `sys/random.h` in `configure.cmake`, just in case.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,218 @@
-            if (fd != -1)
-            {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
```
Why did this get removed? Now parallel open leaks a file descriptor.

lpereira replied:
Not an issue any longer: the /dev/urandom is opened just once in a constructor function.  There's no need to use an atomic CAS like before.

bartonjs replied:
I see that this is now only called from a library load constructor.  So a comment explaining that would be good.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,218 @@
+
+    if (!sInitializedMRand)
+    {
+        // FIXME: There's not enough entropy in time(NULL) to seed the PRNG.
```
This comment should either be removed, or an issue should be opened for it and this should track back.

e.g. // TODO(37906): There's ....

lpereira replied:
Right.  I'm just learning the best practices around here :)  Will open an issue to track it and edit the comment.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,218 @@
+}
+#endif // HAVE_SYS_GETRANDOM
 
-    // always xor srand48 over the whole buffer to get some randomness
```
Why did you get rid of this behavior?

lpereira replied:
Because it's better to trust urandom, that was properly seeded (and is the product of various other sources of random data already), rather than using another PRNG that has been seeded with `time(NULL)` to flip some of those bits.

The previous behavior not only did unnecessary work, but was also questionable.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,218 @@
+        FillWithMrand(buffer, bufferLength);
+    }
+}
 #endif // HAVE_ARC4RANDOM_BUF
```
This is now ending !HAVE_ARC4RANDOM_BUF.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,218 @@
+}
 #endif // HAVE_ARC4RANDOM_BUF
+
+#if HAVE_ARC4RANDOM_BUF
```
Should this just be #else?

bartonjs commented on src/Native/Unix/configure.cmake:
```diff
@@ -193,6 +193,11 @@ check_symbol_exists(
+check_symbol_exists(
+    SYS_getrandom
+    "sys/syscall.h;sys/random.h"
+    HAVE_SYS_GETRANDOM)
```
Right now this is OK because we build the portable asset on a Linux distro that has this defined.  But if we were to change to build portable from CentOS7, RHEL6, or RHEL7 then this change would disappear.

I wonder if we should just locally replicate the syscall ID and definition of GRND_NONBLOCK to maintain flexibility on our build lab.  (With comments to explain why)

lpereira replied:
I'm fine either way, but the existence of `sys/random.h` is tested here because of a comment @am11 made.  It was using syscall(2) directly before.

bartonjs replied:
Well, I think that the PR at the time had both the `check_symbol_exists` for `SYS_getrandom` and a manual define for GRND_NONBLOCK.  I like them being treated the same, but my comment is to question if there should not be a `check_symbol_exists` / `HAVE_SYS_GETRANDOM`, but rather just

```C
#if LINUX

#ifdef __NR_getrandom
static_assert(__NR_getrandom == 278);
#endif

// Equivalent to SYS_getrandom, but sidesteps the forward/re-definition since there are both
// pre-3.14 and post-3.14 kernels in the distro support matrix.
#define PAL_getrandom 278
```

Since the fallback for it not existing (if we changed our setup) is to just use urandom, perhaps this is good (and better than a forward define, since it avoids assuming the syscall ID on any future architectures) as-is.

Merely an exploratory question.

lpereira replied:
Not really a good idea to hardcode syscall values: they're arch-specific.

lpereira replied:
In any case, yes, doing some conditional platform-specific compilation is fine by me.  Kind of a po-tay-to/po-tah-to situation; I'm fine either way.

lpereira replied:
Resolved.

jkotas replied:
Can you please verify that the `getrandom` syscall path is going to be present in the binaries that we ship that are built on CentOS7?

> Not really a good idea to hardcode syscall numbers: they're arch-specific

Yes, but it won't be the first place we do this: https://github.com/dotnet/coreclr/blob/0463767cd7362a0044c4c7af0feda5bd9ad94804/src/gc/unix/gcenv.unix.cpp#L46 . I do not think there is a better way to do the light up for newer kernels on Linux.

cc @janvorli

lpereira replied:
This has been resolved.  `syscall()` is used now instead of `getrandom()`.

lpereira commented:
Thanks everybody for the review -- I force-pushed a new version addressing most of the comments.

jkotas commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,219 @@
+
+static RandomFillFunction g_randomFillFunction = FillWithMrand;
+
+__attribute__((constructor(65535)))
```
The static constructors are problematic. We try to avoid them:
- Ordering problems when you get more of them
- Performance overhead of doing unnecessary work during startup
- They are not standard C/C++.

lpereira replied:
Resolved.

lpereira commented:
Force-pushed a new version with the requested changes:

- No static constructors, but atomic CAS is back
- No changes to the build system to determine availability of getrandom() -- rely on `SYS_getrandom` to be defined on Linux
- Buffer is always XOR-ed with result from lrand48()

maryamariyan commented:
/azp run corefx-ci

maryamariyan commented:
There was failure on Windows x86_Release, so I am retriggering the CI.

@lpereira thank you for the PR.
Some questions for you:

- There seems to be a few unresolved comments remaining on this PR. Are you addressing them in a later PR? 
- I see you linked this to another PR (https://github.com/dotnet/corefx/issues/38034). Is that one a prerequisite for this PR? If so, we could close this and reopen after that has been merged.

We are planning to reduce the number of open PRs in corefx and only keep the active ones open. 
If you think this will not be merged soon, then it is best if we comment on what are the actionable items here and then perhaps close and reopen when it is ready. 

Thanks again and please let me know how you see the timeline and actionable items for the open PRs.

lpereira commented:
@maryamariyan Comments were resolved already, just not marked as resolved here on GitHub.  I've marked them as such here.

I don't think there are any impediments to having this merged (with the exception of the Windows x86_Release, which we'll know soon enough if there's something else to be done).

lpereira commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,258 @@
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+        newFd = open("/dev/urandom", O_RDONLY);
+        fcntl(fd, F_SETFD, FD_CLOEXEC);
```
```suggestion
        fcntl(newFd, F_SETFD, FD_CLOEXEC);
```

jkotas commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,258 @@
+    {
+        bool lRandUninitialized = false;
+
+        if (__atomic_compare_exchange_n(&sInitializedLrand, &lRandUninitialized, true, false,
```
This sets the flag first and then does the initialization. Shouldn't it be the other way around?

bartonjs replied:
I think that it's trying to avoid the side effect of calling srand a second time.

Seems like it does need some sort of wait for the race losers so they don't call lrand48 before it's seeded.

```C

if (!sInitializedLrand)
{
    static bool sShouldIDoIt;
    bool lShouldIDoIt = false;

    if (atomic exchange(shouldIDoIt))
    {
        srand48(time(NULL));
        sInitializedLrand = true;
    }
    else
    {
        while (!sInitializedLrand)
        {
            sleep(1);
        }
    }
}
```

?

lpereira replied:
Indeed, nice catch.  Will re-push a fix shortly after this builds and I'm able to test by intentionally introducing a race (adding a sleep() in the case where the CAS succeeds).

bartonjs reviewed (changes requested):
The race condition between srand48 and potential parallel calls to lrand48 needs to be addressed (potentially including explaining why it isn't a problem)

safern commented:
@lpereira what's the current state of this PR, do you have any plans on addressing feedback soon?

lpereira commented:
> @lpereira what's the current state of this PR, do you have any plans on addressing feedback soon?

Feedback will be addressed shortly.  Notifications for this item ended up falling through the cracks of my email client.

lpereira commented:
Rebased on top of current master and pushed addressing comments.

maryamariyan commented:
@bartonjs @joperezr this PR seems to be ready to review again.

bartonjs commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,269 @@
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
+                sched_yield();
```
Is it possible that a normal priority thread won the compare exchange, but a high priority thread gets stuck in here?  Will that just deadlock, or will the high priority thread eventually yield anyways? (the description of sched_yield suggests that a high priority thread just ignores the yield call).

bartonjs commented:
The rebase (well, the force-push) made it difficult to review what was different, but it looks like the srand race condition has been alleviated (modulo one thread priority question). Any other feedback, @jkotas?

jkotas commented:
> Any other feedback, @jkotas?

This: https://github.com/dotnet/corefx/pull/37906#discussion_r309979158

stephentoub commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,269 @@
-        {
-            int fd;
+        return fd;
+    }
```
There are no issues with the thread-safety around the reads of this field? e.g. is there any situation where the compiler/hardware might introduce a read such that it effectively became `int a = fd; if (fd >= 0) return a;`?

stephentoub commented on src/Native/Unix/System.Native/pal_random.c:
```diff
@@ -2,109 +2,269 @@
+    do
+    {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
```
How was this not causing problems for us before?  Does the value of O_CLOEXEC happen to map to the correct bits that make this readable by the current user?

lpereira replied:
Because the third parameter is ignored if the second doesn't contain `O_CREAT`.

stephentoub commented:
@lpereira, are you planning to continue working on this?

lpereira commented:
> @lpereira, are you planning to continue working on this?

I am.  I'll address these comments this week; been busy with other things.

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

joperezr commented:
Hey @lpereira still planning on finishing this? The cutoff for the migration is tomorrow.

maryamariyan commented:
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.
